### DISCUSSION TITLE
Why does prisma migrate dev want to reset my db?

### DISCUSSION THREAD

**Discussion by jsw** [prisma/prisma#18076] [Q&A] [answered]
I'm trying to create a small migration with a column name change. I'm  trying to understand why `prisma migrate diff` recognizes the simple change, but `prisma migrate dev` wants to reset the db. I am using MySQL and already have an initial migration, and have not made any other changes to `prisma.schema` or directly to the database.

```
$ npx prisma migrate diff --from-url mysql://root:root@localhost:3306/api --to-schema-datamodel prisma/schema.prisma  

[*] Changed the `respondent` table
  [-] Removed column `group`
  [+] Added column `groupName`
```

```
$ npx prisma migrate dev --name group_to_groupname
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "api" at "localhost:3306"

Drift detected: Your database schema is not in sync with your migration history.

The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.

It should be understood as the set of changes to get from the expected schema to the actual schema.

If you are running this the first time on an existing database, please make sure to read this documentation page:
https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/troubleshooting-development

[+] Added tables
  - _CollectorToSurvey
  - collector
  - organization
  - respondent
  - survey
  - user

[*] Changed the `_CollectorToSurvey` table
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `collector` table
  [+] Added foreign key on columns (organizationId)

[*] Changed the `respondent` table
  [+] Added foreign key on columns (collectorId)
  [+] Added foreign key on columns (organizationId)
  [+] Added foreign key on columns (parentId)
  [+] Added foreign key on columns (surveyId)
  [+] Added foreign key on columns (userId)

[*] Changed the `survey` table
  [+] Added foreign key on columns (organizationId)

[*] Changed the `user` table
  [+] Added foreign key on columns (organizationId)

✔ We need to reset the MySQL database "api" at "localhost:3306"
Do you want to continue? All data will be lost. … no
```

**Comment by Karl-EdwardJeanMehu** [comment, marked as answer]
@jsw You changed the column name in the schema correct? Any changes whatsoever to the schema, or indirectly (through introspection) will trigger a discrepancy check with the existing migration history. New migrations will be created and applied to and a new client will be generated to reflect those changes. [Prisma Migration in Development and Production](https://www.prisma.io/docs/concepts/components/prisma-migrate/migrate-development-production#create-and-apply-migrations)

**Reply by jsw** [reply]
@Karl-EdwardFPJeanMehu I changed the column name in prisma.schema and have not altered the database since the initial migration was performed. What I don't follow is why there is schema drift, if that's all I did. The linked page states that it will prompt a reset in the following conditions, neither of which I believe should be true in my case. You'll notice above that the `diff` command contains the expected diff, but the `migrate` command detects that everything is off. Is this expected for every change in dev?

> Migration history conflicts caused by [modified or missing migrations](https://www.prisma.io/docs/concepts/components/prisma-migrate#do-not-edit-or-delete-migrations-that-have-been-applied)
    The database schema has drifted away from the end-state of the migration history

**Reply by janpio** [reply]
Your `diff` command compares the database to the schema file.
The drift detection compares your migration history (so the files in your `prisma/migrations` folder) to the current state of the database.

How did you create your migration history?
How did you create the database and schema file?

Most probably the soluion to your problem is to baseline your migration history: https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/baselining

**Reply by jsw** [reply]
This was several months ago, but looking at my history, it looks like after prototyping via manual changes to my `prisma.schema` and `prisma db push` I eventually did the following in an attempt to right everything and switch to using `migrate`:

* `prisma db pull`
* `prisma db generate`
* `prisma migrate dev --name initial-migration --create-only`
* `prisma migrate dev`

This overwrote my `prisma.schema`, reset my db, and created the `prisma/migrations` folder. In CI/CD I have a `prisma migrate deploy` step against RDS MySQL, which has been working as expected. Now, several months later, I wish to change a column name. In case it's relevant, I'm running mysql 8 via docker compose. 

Here is the relevant parts of the initial migration.sql for a particular table followed by the output of `show create table`

```
-- CreateTable
CREATE TABLE `user` (
    `id` VARCHAR(191) NOT NULL,
    `organizationId` VARCHAR(191) NOT NULL,
    `authId` VARCHAR(191) NOT NULL,
    `invitedAt` DATETIME(3) NULL,
    `metadata` JSON NULL,
    `createdAt` DATETIME(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
    `updatedAt` DATETIME(3) NOT NULL,
    `deletedAt` DATETIME(3) NULL,

    UNIQUE INDEX `user_authId_key`(`authId`),
    INDEX `user_organizationId_fkey`(`organizationId`),
    PRIMARY KEY (`id`)
) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;

-- AddForeignKey
ALTER TABLE `user` ADD CONSTRAINT `user_organizationId_fkey` FOREIGN KEY (`organizationId`) REFERENCES `organization`(`id`) ON DELETE RESTRICT ON UPDATE CASCADE;
```

```
mysql> show create table user\G
*************************** 1. row ***************************
       Table: user
Create Table: CREATE TABLE `user` (
  `id` varchar(191) COLLATE utf8mb4_unicode_ci NOT NULL,
  `organizationId` varchar(191) COLLATE utf8mb4_unicode_ci NOT NULL,
  `authId` varchar(191) COLLATE utf8mb4_unicode_ci NOT NULL,
  `invitedAt` datetime(3) DEFAULT NULL,
  `metadata` json DEFAULT NULL,
  `createdAt` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
  `updatedAt` datetime(3) NOT NULL,
  `deletedAt` datetime(3) DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `user_authId_key` (`authId`),
  KEY `user_organizationId_fkey` (`organizationId`),
  CONSTRAINT `user_organizationId_fkey` FOREIGN KEY (`organizationId`) REFERENCES `organization` (`id`) ON DELETE RESTRICT ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
1 row in set (0.01 sec)
```

**Reply by janpio** [reply]
Those are different tables 😆 

But your description sounds valid. Is your local current database at the correct state (via `migrate deploy` or `migrate dev`) before you make the change to the schema file and try to generate a new migration? Anything blocking you from just resetting the local development database?

**Reply by jsw** [reply]
Oops, I edited my comment to use the same tables. Here is the output of `migrate deploy` and `migrate deploy` before any changes are made to `prisma.schema`. I could reset my local database (I am guessing the other developers will be forced to do the same), but am trying to understand what's wrong with my workflow, and the smoother it runs locally the more confident I am in deploying changes to the shared databases via `migrate deploy`.

```
➜ npx prisma migrate deploy
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "api" at "localhost:3306"

1 migration found in prisma/migrations

Error: P3005

The database schema is not empty. Read more about how to baseline an existing production database: https://pris.ly/d/migrate-baseline

➜  npx prisma migrate dev
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "api" at "localhost:3306"

Drift detected: Your database schema is not in sync with your migration history.

The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.

It should be understood as the set of changes to get from the expected schema to the actual schema.

If you are running this the first time on an existing database, please make sure to read this documentation page:
https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/troubleshooting-development

[+] Added tables
  - _CollectorToSurvey
  - collector
  - organization
  - respondent
  - survey
  - user

[*] Changed the `_CollectorToSurvey` table
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `collector` table
  [+] Added foreign key on columns (organizationId)

[*] Changed the `respondent` table
  [+] Added foreign key on columns (collectorId)
  [+] Added foreign key on columns (organizationId)
  [+] Added foreign key on columns (parentId)
  [+] Added foreign key on columns (surveyId)
  [+] Added foreign key on columns (userId)

[*] Changed the `survey` table
  [+] Added foreign key on columns (organizationId)

[*] Changed the `user` table
  [+] Added foreign key on columns (organizationId)

✔ We need to reset the MySQL database "api" at "localhost:3306"
Do you want to continue? All data will be lost. … no
```

**Reply by janpio** [reply]
The tables look identical now.

Both `migrate deploy` and `migrate dev` complaining indicates to me that something is off with this development database.

- For `migrate deploy` I could theorize that it is expecting the database to be empty because it does not have any entries in the `_prisma_migrations` table (or the table does not even exist, so is created).
- For `migrate dev` it seems to be convinced that the migration history (`migration.sql` files) and the database state do not match.

How did you create this development database initially?

**Reply by jsw** [reply]
I believe it was initially created via series of `db push` and then reset via `migrate dev` with the initial migration.

**Reply by janpio** [reply]
I suggest you craft a `migrate diff` command that compares your migration history (files) to your database, probably something like `npx prisma migrate diff --from-migrations ./prisma/migrations --to-schema-datasource ./prisma/schema.prisma`. That should run similar code than what `migrate dev` runs and tell you what it sees as the difference between the migration files and the database you are comparing to. According to the SQL you posted above, that should not detect any differences for the `user` table for example. If it does, something is off or weird.

**Reply by jsw** [reply]
This command requires a shadow database, so I manually created a shadow database via `create database shadow` and then ran

```
npx prisma migrate diff --from-migrations ./prisma/migrations --to-schema-datasource ./prisma/schema.prisma --shadow-database-url mysql://root:root@localhost:3306/shadow
No difference detected.
```

And just to confirm things are still off I ran `npx prisma migrate dev` and got the "Drift detected" as before.

**Reply by janpio** [reply]
That is _weird_.

What version of Prisma are you using?
Is that list of 6 tables in the drift output all your models (actually probably only 5 but one m:n relation)?
Are you using a `relationMode` property?

Are you running your local MySQL in Docker, so you could potentially share it with us? I fear the best way forward would be to have a look ourselves.

**Reply by jsw** [reply]
Prisma 4.10.1.
Correct, the drift output is all 5 of my models plus the m:n relationship.
No `relationMode` property.
I am running mysql in docker.
Happy to privately share my prisma folder and docker container with you.
Let me know how best to package it up and share. 

```
prisma                  : 4.10.1
@prisma/client          : 4.10.1
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine aead147aa326ccb985dcfed5b065b4fdabd44b19 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli aead147aa326ccb985dcfed5b065b4fdabd44b19 (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.10.1-1.80b351cc7c06d352abe81be19b8a89e9c6b7c110
Default Engines Hash    : aead147aa326ccb985dcfed5b065b4fdabd44b19
Studio                  : 0.481.0
```

**Reply by janpio** [reply]
I am reachable at lastname@prisma.io - I can accept attachments or links to Dropbox, Google Drive - whatever. You can also put it into a (private) GitHub repo and invite my user - whatever works best for you.

**Reply by janpio** [reply]
Ok, figured it out.

Your `api` database does not have a `_prisma_migrations` table where the applied migrations are recorded. `migrate dev` uses that to figure out which of the migration files it actually should apply (in this case: none) and then compares that state to your database - and that would mean a lot of drift indeed.

The output of `migrate dev` is not super helpful here as it does not mention that it uses the migration table to figure out which migrations to apply, and also does not in any way indicate that it is missing that table or that it is only applying specific migration files.

So it seems your dev database `api` might just have been created via `db push` instead of `migrate dev` or `migrate deploy`, or the `_prisma_migrations` table got lost along the way somehow. On a database that has the actual migration history, `migrate dev` should successfully create an empty migration.

To fix: `npx prisma migrate resolve --applied 20221218174519_initial_migration` This will create the migration table `_prisma_migrations` and also mark the existing one migration as applied. Then `migrate dev` will run the correct comparison, and will not detect any drift any more.

**Reply by jsw** [reply]
Nice! That makes sense. It may be the case that I had run a `db push` at some point as a convenient way to "clear" the database, not realizing it would also drop the `_prisma_migrations` table. Thanks much for your assistance on this. I now have more confidence in the correct workflow.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
